validator_client/src/block_service.rs

```diff
@@ -320,87 +320,217 @@ impl<T: SlotClock + 'static, E: EthSpec> BlockService<T, E> {
+                .validator_store
+                .sign_block(*validator_pubkey, block, current_slot)
+                .await
+                .map(SignedBlock::Blinded),
+        };
+
+        let signed_block = match res {
+            Ok(block) => block,
+            Err(ValidatorStoreError::UnknownPubkey(pubkey)) => {
+                // A pubkey can be missing when a validator was recently removed
+                // via the API.
+                warn!(
+                    log,
+                    "Missing pubkey for block";
+                    "info" => "a validator may have recently been removed from this VC",
+                    "pubkey" => ?pubkey,
+                    "slot" => ?slot
+                );
+                return Ok(());
+            }
+            Err(e) => {
+                return Err(BlockError::Recoverable(format!(
+                    "Unable to sign block: {:?}",
+                    e
+                )))
+            }
+        };
+
+        let signing_time_ms =
+            Duration::from_secs_f64(signing_timer.map_or(0.0, |t| t.stop_and_record())).as_millis();
+
+        info!(
+            log,
+            "Publishing signed block";
+            "slot" => slot.as_u64(),
+            "signing_time_ms" => signing_time_ms,
+        );
+
+        // Publish block with first available beacon node.
+        //
+        // Try the proposer nodes first, since we've likely gone to efforts to
+        // protect them from DoS attacks and they're most likely to successfully
+        // publish a block.
+        proposer_fallback
+            .request_proposers_first(
+                RequireSynced::No,
+                OfflineOnFailure::Yes,
+                |beacon_node| async {
+                    self.publish_signed_block_contents(&signed_block, beacon_node)
+                        .await
+                },
+            )
+            .await?;
+
+        info!(
+            log,
+            "Successfully published block";
+            "block_type" => ?signed_block.block_type(),
+            "deposits" => signed_block.num_deposits(),
+            "attestations" => signed_block.num_attestations(),
+            "graffiti" => ?graffiti.map(|g| g.as_utf8_lossy()),
+            "slot" => signed_block.slot().as_u64(),
+        );
+        Ok(())
+    }
+
+    async fn publish_block_v3(
+        self,
         slot: Slot,
         validator_pubkey: PublicKeyBytes,
-        builder_proposal: bool,
+        builder_boost_factor: Option<u64>,
     ) -> Result<(), BlockError> {
         let log = self.context.log();
         let _timer =
             metrics::start_timer_vec(&metrics::BLOCK_SERVICE_TIMES, &[metrics::BEACON_BLOCK]);
 
         let current_slot = self.slot_clock.now().ok_or_else(|| {
-            BlockError::Recoverable("Unable to determine current slot from clock".to_string())
+            BlockError::Irrecoverable("Unable to determine current slot from clock".to_string())
```

Comment:
I don't think this should be changed to irrecoverable? We haven't signed anything, which is what makes a failure irrecoverable.

Reply:
Ah I see! 
I suggested changing this to `Irrecoverable` earlier because for the block V3 endpoints we don't do any retries. But if this it what distinguishes between `Irrecoverable` and `Recoverable` it might be worth adding a comment on the enum.

Reply:
Great idea, I'd assumed that comment already existed :sweat_smile:

Reply:
This line no longer exists since https://github.com/sigp/lighthouse/pull/4813/commits/32cad117601a47b52807f77ad418515f252f1a88, we now use the `slot` function param instead of reading from the `slot_clock` again.